\section{Experiments}

We have implemented the relative suffix tree in C++, extending the
old relative FM-index implementation.\footnote{The current implementation is available
at \url{https://github.com/jltsiren/relative-fm}.} The implementation is based
on the \emph{Succinct Data Structure Library (SDSL) 2.0}~\cite{Gog2014b}. Some
parts of the implementation have been parallelized using \emph{OpenMP} and the
\emph{libstdc++ parallel mode}.

As our reference \CSA{}, we used the \emph{succinct suffix array} (\SSA{})
\cite{Ferragina2007a,Maekinen2005} implemented using SDSL components. Our
implementation is very similar to \texttt{csa\_wt} in SDSL, but we needed
better access to the internals than what the SDSL interface
provides. \SSA{} encodes the Burrows-Wheeler transform as a \emph{Huffman-shaped
wavelet tree}, combining fast queries with size close to the
\emph{order\nobreakdash-$0$ empirical entropy}. This makes it the
index of choice for DNA sequences \cite{Ferragina2009a}. In addition to
the plain \SSA{} with uncompressed bitvectors, we also used \SSArrr{} with
entropy-compressed bitvectors \cite{Raman2007} to highlight the
the time-space trade-offs achieved with better compression

We sampled \SA{} in suffix order and \ISA{} in text order. In \SSA, the sample
intervals were $17$ for \SA{} and $64$ for \ISA. In \RFM, we used sample
interval $257$ for \SA{} and $512$ for \ISA{} to handle the regions that do
not exist in the reference. The sample intervals for suffix order sampling
were primes due to the long runs of character $N$ in the assembled genomes. If
the number of long runs of character $N$ in the indexed sequence is even, the
lexicographic ranks of almost all suffixes in half of the runs are odd, and
those runs are almost completely unsampled. This can be avoided by making the
sample interval and the number of runs \emph{relatively prime}.

The experiments were done on a system with two 16\nobreakdash-core AMD Opteron 6378 processors and 256~GB of memory. The system was running Ubuntu 12.04 with Linux kernel 3.2.0. We compiled all code with g++ version~4.9.2. We allowed index construction to use multiple threads, while confining the query benchmarks to a single thread. As AMD Opteron uses a \emph{non-uniform memory access} architecture, accessing local memory controlled by the same physical CPU is faster than accessing remote memory controlled by another CPU. In order to ensure that all data structures are in local memory, we set the CPU affinity of the query benchmarks with the \texttt{taskset} utility.

As our target sequence, we used the \emph{maternal haplotypes} of the
\emph{1000 Genomes Project individual NA12878}~\cite{Rozowsky2011}. As the
reference sequence, we used the 1000 Genomes Project version of the \emph{GRCh37
assembly} of the \emph{human reference
genome}.\footnote{\url{ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/}}
Because NA12878 is female, we also created a reference sequence without
chromosome~Y.

In the following, a basic FM-index is an index supporting only \find{}
queries, while a full index also supports \locate{} and \extract{} queries.

\subsection{Indexes and their sizes}

Table~\ref{table:construction} lists the resource requirements for building
the relative indexes, assuming that we have already built the corresponding
non-relative structures for the sequences. As a comparison, building an
FM-index for a human genome typically takes 16--17 minutes and 25--26~GB
of memory. While the construction of the basic \RFM{} index is
highly optimized, the other construction algorithms are just the first
implementations. Building the optional \rselect{} structures takes 4 minutes
using two threads and around 730~megabytes ($\abs{R} + \abs{S}$ bits) of
working space in addition to \RFM{} and \rselect.

\begin{table*}
\caption{Sequence lengths and resources used by index construction for NA12878
relative to the human reference genome with and without chromosome~Y. Approx
and Inv denote the approximate \LCS{} and the bwt-invariant subsequence.
Sequence lengths are in millions of base pairs, while construction resources
are in minutes of wall clock time and gigabytes of
memory.}\label{table:construction}
\setlength{\extrarowheight}{2pt}
\setlength{\tabcolsep}{3pt}
\begin{center}
\begin{tabular}{c|cccc|cc|cc|cc}
\hline
 &
\multicolumn{4}{c|}{\textbf{Sequence length}} &
\multicolumn{2}{c|}{\textbf{\RFM{} (basic)}} &
\multicolumn{2}{c|}{\textbf{\RFM{} (full)}} &
\multicolumn{2}{c}{\textbf{\RCST}} \\
\textbf{ChrY} &
\textbf{Reference} & \textbf{Target} & \textbf{Approx} & \textbf{Inv} &
\textbf{Time} & \textbf{Memory} &
\textbf{Time} & \textbf{Memory} &
\textbf{Time} & \textbf{Memory} \\
\hline
yes & 3096M & 3036M & 2992M & 2980M & 1.42 min & 4.41 GB & 175 min & 84.0 GB &
629 min & 141 GB \\
no  & 3036M & 3036M & 2991M & 2980M & 1.33 min & 4.38 GB & 173 min & 82.6 GB &
593 min & 142 GB \\
\hline
\end{tabular}
\end{center}
\end{table*}

The sizes of the final indexes are listed in Table~\ref{table:indexes}. The
full \RFM{} is over twice the size of the basic index, but still
3.3\nobreakdash––3.7~times smaller than the full \SSArrr{} and
4.6\nobreakdash––5.3~times smaller than the full \SSA. The \RLCP{} array is
2.7~times larger than the \RFM{} index with the full human reference and
1.5~times larger with the female reference. Hence having a separate female
reference is worthwhile, if there are more than a few female genomes among
the target sequences. The optional \rselect{} structure is almost as large
as the basic \RFM{} index.

\begin{table*}
\caption{Various indexes for NA12878 relative to the human reference genome
with and without chromosome~Y. The total for \RCST{} includes the full \RFM.
Index sizes are in megabytes and in bits per character.}\label{table:indexes}
\setlength{\extrarowheight}{2pt}
\setlength{\tabcolsep}{3pt}
\begin{center}
\begin{tabular}{c|cc|cc|cc|ccc}
\hline
 &
\multicolumn{2}{c|}{\textbf{\SSA}} &
\multicolumn{2}{c|}{\textbf{\SSArrr}} &
\multicolumn{2}{c|}{\textbf{\RFM}} &
\multicolumn{3}{c}{\textbf{\RCST}} \\
\textbf{ChrY} &
\textbf{Basic} & \textbf{Full} &
\textbf{Basic} & \textbf{Full} &
\textbf{Basic} & \textbf{Full} &
\textbf{\RLCP} & \textbf{Total} & \textbf{\rselect} \\
\hline
\multirow{2}{*}{yes}
&  1248 MB &  2110 MB &   636 MB &  1498 MB &   225 MB &   456 MB &  1233 MB &  1689 MB &   190 MB \\
& 3.45 bpc & 5.83 bpc & 1.76 bpc & 4.14 bpc & 0.62 bpc & 1.26 bpc & 3.41 bpc & 4.67 bpc & 0.52 bpc \\
\hline
\multirow{2}{*}{no}
&  1248 MB &  2110 MB &   636 MB &  1498 MB &   186 MB &   400 MB &   597 MB &   997 MB &   163 MB \\
& 3.45 bpc & 5.83 bpc & 1.76 bpc & 4.14 bpc & 0.51 bpc & 1.11 bpc & 1.65 bpc & 2.75 bpc & 0.45 bpc \\
\hline
\end{tabular}
\end{center}
\end{table*}

Table~\ref{table:rfm components} lists
the sizes of the individual components of the relative FM-index.
Including the chromosome~Y in the reference increases the sizes
of almost all relative components, with the exception of $\mCS(\mBWT_{S})$ and
$\mAli(R,S)$. In the first case, the common subsequence still covers
approximately the same positions in $\mBWT_{S}$ as before. In the second case,
chromosome~Y appears in bitvector $B_{R}$ as a long run of \zerobit{}s, which
compresses well. The components of a full \RFM{} index are larger
than the corresponding components of a basic \RFM{} index, because the
bwt-invariant subsequence is shorter than the approximate longest
common subsequence (see Table~\ref{table:construction}).

\begin{table*}
\caption{Breakdown of component sizes in the \RFM{} index for NA12878 relative
to the human reference genome with and without chromosome~Y in bits per
character.}\label{table:rfm components}
\setlength{\extrarowheight}{2pt}
\setlength{\tabcolsep}{3pt}
\begin{center}
\begin{tabular}{c|cc|cc}
\hline
 & \multicolumn{2}{c|}{\textbf{Basic \RFM}} & \multicolumn{2}{c}{\textbf{Full
\RFM}} \\
\textbf{ChrY}                &      \textbf{yes} &       \textbf{no} &      \textbf{yes} &       \textbf{no} \\
\hline
\textbf{\RFM}                & \textbf{0.62 bpc} & \textbf{0.51 bpc} & \textbf{1.26 bpc} & \textbf{1.11 bpc} \\
$\mCS(\mBWT_{R})$            &          0.12 bpc &          0.05 bpc &          0.14 bpc &          0.06 bpc \\
$\mCS(\mBWT_{S})$            &          0.05 bpc &          0.05 bpc &          0.06 bpc &          0.06 bpc \\
$\mAli(\mBWT_{R},\mBWT_{S})$ &          0.45 bpc &          0.42 bpc &          0.52 bpc &          0.45 bpc \\
$\mAli(R,S)$                 &                -- &                -- &          0.35 bpc &          0.35 bpc \\
\SA{} samples                &                -- &                -- &          0.12 bpc &          0.12 bpc \\
\ISA{} samples               &                -- &                -- &          0.06 bpc &          0.06 bpc \\
\hline
\end{tabular}
\end{center}
\end{table*}

The size breakdown of the \RLCP{} array can be seen in Table~\ref{table:rlcp components}.
Phrase pointers and phrase lengths take space proportional to the number of phrases. As
there are more mismatches between the copied substrings with the full human reference
than with the female reference, the absolute \LCP{} values take a larger proportion of the
total space with the full reference. Shorter phrase length increases the likelihood that
the minimal \LCP{} value in a phrase is a large value, increasing the size of the minima tree.

\begin{table*}
\caption{Breakdown of component sizes in the \RLCP{} array for NA12878 relative
to the human reference genome with and without chromosome~Y. The number of phrases,
average phrase length, and the component sizes in bits per character. ``Parse''
contains $W_{r}$ and $W_{\ell}$, ``Literals'' contains $W_{c}$ and $L$, and ``Tree''
contains $M_{\mLCP}$ and $M_{L}$.}\label{table:rlcp components}
\setlength{\extrarowheight}{2pt}
\setlength{\tabcolsep}{3pt}
\begin{center}
\begin{tabular}{c|cc|ccc|c}
\hline
\textbf{ChrY} & \textbf{Phrases} & \textbf{Length} & \textbf{Parse} & \textbf{Literals} & \textbf{Tree} & \textbf{Total} \\
\hline
yes           &      128 million &            23.6 &      1.35 bpc &          1.54 bpc &      0.52 bpc &        3.41 bpc \\
no            &       94 million &            32.3 &      0.97 bpc &          0.41 bpc &      0.27 bpc &        1.65 bpc \\
\hline
\end{tabular}
\end{center}
\end{table*}

\subsection{Query times}

Average query times for the basic operations can be seen in Tables~\ref{table:rfm
queries} and~\ref{table:rlcp queries}. The results for \LF{} and \Psiop{} queries
in the full FM-indexes are similar to the earlier ones with basic indexes
\cite{Boucher2015}. Random access to the \RLCP{} array is about 30~times
slower than to the \LCP{} array, while sequential access is 10~times slower.
The \nsv, \psv, and \rmq{} queries are comparable to 1\nobreakdash--2~random
accesses to the \RLCP{} array.

\begin{table*}
\caption{Average query times in microseconds for 10~million random queries in the full
\SSA, the full \SSArrr, and the full \RFM{} for NA12878 relative to
the human reference genome with and without chromosome~Y.}\label{table:rfm
queries}
\setlength{\extrarowheight}{2pt}
\setlength{\tabcolsep}{3pt}
\begin{center}
\begin{tabular}{c|cc|cc|cc|c}
\hline
& \multicolumn{2}{c|}{\textbf{\SSA}} & \multicolumn{2}{c|}{\textbf{\SSArrr}} & \multicolumn{2}{c|}{\textbf{\RFM}} & \textbf{\rselect} \\
\textbf{ChrY} & \textbf{\LF} & \textbf{\Psiop} & \textbf{\LF} & \textbf{\Psiop} & \textbf{\LF} & \textbf{\Psiop} & \textbf{\Psiop} \\
\hline
yes           &   0.328 \mus &      1.048 \mus &   1.989 \mus &      2.709 \mus &   3.054 \mus &     43.095 \mus &      5.196 \mus \\
no            &   0.327 \mus &      1.047 \mus &   1.988 \mus &      2.707 \mus &   2.894 \mus &     40.478 \mus &      5.001 \mus \\
\hline
\end{tabular}
\end{center}
\end{table*}

\begin{table*}
\caption{Query times in microseconds in the \LCP{} array (\slarray) and the
\RLCP{} array for NA12878 relative to the human reference genome with and without
chromosome~Y. For the random queries, the query times are averages over
100~million queries. The range lengths for the \rmq{} queries were $16^{k}$ (for
$k \ge 1$) with probability $0.5^{k}$. For sequential access, we list the average
time per position for scanning the entire array.}\label{table:rlcp
queries}
\setlength{\extrarowheight}{2pt}
\setlength{\tabcolsep}{3pt}
\begin{center}
\begin{tabular}{c|cc|ccccc}
\hline
& \multicolumn{2}{c|}{\textbf{\LCP{} array}} & \multicolumn{5}{c}{\textbf{\RLCP{} array}} \\
\textbf{ChrY} & \textbf{Random} & \textbf{Sequential} & \textbf{Random} & \textbf{Sequential} & \textbf{\nsv} & \textbf{\psv} & \textbf{\rmq} \\
\hline
yes           &      0.054 \mus &          0.002 \mus &      1.580 \mus &          0.024 \mus &    1.909 \mus &    1.899 \mus &    2.985 \mus \\
no            &      0.054 \mus &          0.002 \mus &      1.480 \mus &          0.017 \mus &    1.834 \mus &    1.788 \mus &    3.078 \mus \\
\hline
\end{tabular}
\end{center}
\end{table*}

We also tested the \locate{} performance of the full \RFM{} index, and
compared it to \SSA{} and \SSArrr. We built the indexes with \SA{} sample
intervals $7$, $17$, $31$, $61$, and $127$, using the reference without
chromosome~Y for \RFM.\footnote{With \RFM, the sample intervals apply
to the reference \SSA.} The \ISA{} sample interval was the
maximum of $64$ and the \SA{} sample interval. We extracted 2~million
random patterns of length $32$, consisting of characters $ACGT$, from
the target sequence, and measured the total time taken by \find{} and
\locate{} queries. The results can be seen in
Figure~\ref{fig:locate}. While \SSA{} and \SSArrr{} query times were
proportional to the sample interval, \RFM{} used 5.4\nobreakdash--7.6~microseconds
per occurrence more than \SSA{}, resulting in slower growth in query times.
In particular, \RFM{} with reference sample interval $31$ was faster than
\SSA{} with sample interval $61$.

\begin{figure*}
\begin{center}
\includegraphics{locate_bars.pdf}%
\hspace{-0.6in}%
\includegraphics{locate.pdf}
\end{center}
\caption{Average \find{} and \locate{} times in microseconds per occurrence for 2~million patterns
of length $32$ with a total of 255~million occurrences on NA12878 relative to
the human reference genome without chromosome~Y. Left: Query time vs.\ suffix array
sample interval. Right: Query time vs.\ index size in bits per character.}\label{fig:locate}
\end{figure*}

\subsection{Synthetic collections}

In order to determine how the differences between the reference sequence and
the target sequence affect the size of relative structures, we built \RCST{}
for various \emph{synthetic datasets}. We took a 20~MB prefix of the human
reference genome as the reference sequence, and generated 25 target
sequences with every \emph{mutation rate} $p \in \set{0.0001, 0.0003, 0.001, 0.003,
0.01, 0.03, 0.1}$. A total of 90\% of the mutations were single-character
substitutions, while 5\% were insertions and another 5\% deletions. The length of an
insertion or deletion was $k \ge 1$ with probability $0.2 \cdot 0.8^{k-1}$.

The results can be seen in Figure~\ref{fig:synthetic}~(left). The size of the \RLCP{}
array grew quickly with increasing mutation rates, peaking at $p = 0.01$.
At that point, the average length of an \RLZ{} phrase was comparable to what
could be found in the \DLCP{} arrays of unrelated DNA sequences. With even
higher mutation rates, the phrases became slightly longer due to the smaller
average \LCP{} values. The \RFM{} index, on the other hand, remained small
until $p = 0.003$. Afterwards, the index started growing quickly, eventually
overtaking the \RLCP{} array.

\begin{figure*}
\begin{center}
\includegraphics{synth.pdf}%
\hspace{-0.6in}%
\includegraphics{comp.pdf}
\end{center}
\caption{Index size in bits per character vs.\ mutation rate for 25
synthetic sequences relative to a 20~MB reference.}\label{fig:synthetic}
\end{figure*}

We also compared the size of the relative suffix tree to \GCT{} \cite{Navarro2015},
which is essentially a \CSTsada{} for repetitive collections.
While the structures are intended for different purposes, the comparison shows how
much additional space is used for providing access to the suffix trees
of individual datasets. We chose to skip the \CSTnpr{} for repetitive collections
\cite{Abeliuk2013}, as its implementation was not stable enough.

Figure~\ref{fig:synthetic}~(right) shows the sizes of the compressed suffix
trees. The numbers for \RCST{} include individual indexes for each of the 25
target sequences as well as the reference data, while the numbers for \GCT{}
are for a single index containing the 25 sequences. With low mutation rates,
\RCST{} was not much larger than \GCT{}. The size of \RCST{} starts growing
quickly at around $p = 0.001$, while the size of \GCT{} stabilizes at
3\nobreakdash--4~bpc.

\subsection{Suffix tree operations}

In the final set of experiments, we compared the performance of \RCST{} to the
SDSL implementations of various compressed suffix trees. We used the maternal
haplotypes of NA12878 as the target sequence and the human reference genome
without chromosome~Y as the reference sequence. We built \RCST, \CSTsada,
\CSTnpr, and \FCST{} for the target sequence. \CSTsada{} uses \emph{Sadakane's
compressed suffix array} (\CSAsada) \cite{Sadakane2003} as its \CSA, while the
other SDSL implementations use \SSA. We used \PLCP{} as the \LCP{} encoding
with both \CSTsada{} and \CSTnpr{}, and also built \CSTnpr{} with \LCPdac.

We used three algorithms for the performance comparison. The first algorithm is
\emph{preorder traversal} of the suffix tree using SDSL iterators
(\texttt{cst\_dfs\_const\_forward\_iterator}). The iterators use operations
$\mRoot$, $\mLeaf$, $\mParent$, $\mFChild$, and $\mNSibling$, though $\mParent$
queries are rare, as the iterators cache the most recent parent nodes.

The other two algorithms find the \emph{maximal substrings} of the query string
occurring in the indexed text, and report the lexicographic range for each such
substring. This is a key task in common problems such as computing
\emph{matching statistics} \cite{Chang1994} or finding \emph{maximal exact matches}.
The \emph{forward algorithm} uses $\mRoot$, $\mSDepth$, $\mSLink$, $\mChild$,
and $\mLetter$, while the \emph{backward algorithm} \cite{Ohlebusch2010a} uses
$\mLF$, $\mParent$, and $\mSDepth$.

We used the \emph{paternal haplotypes} of chromosome~1 of NA12878 as the
query string in the maximal substrings algorithms. Because some tree operations
in the SDSL compressed suffix trees take time proportional to the depth of the
current node, we truncated the runs of character $N$ in the query string into
a single character. Otherwise searching in the deep subtrees would have
made some SDSL suffix trees much slower than \RCST.

The results can be seen in Table~\ref{table:cst}. \RCST{} was 1.8~times smaller
than \FCST{} and several times smaller than the other compressed suffix trees.
In depth-first traversal, \RCST{} was 4~times slower than \CSTnpr{} and
about 15~times slower than \CSTsada. \FCST{} was orders of magnitude slower,
managing to traverse only 5.3\% of the tree before the run was terminated after
24~hours.

\begin{table*}
\caption{Compressed suffix trees for the maternal haplotypes of NA12878
relative to the human reference genome without chromosome~Y. Component
choices; index size in bits per character; average time in microseconds per
node for preorder traversal; and average time in microseconds per character
for finding maximal substrings shared with the paternal haplotypes of
chromosome~1 of NA12878 using forward and backward algorithms. The figures in
parentheses are estimates based on the progress made in the first 24~hours.}\label{table:cst}
\setlength{\extrarowheight}{2pt}
\setlength{\tabcolsep}{3pt}
\begin{center}
\begin{tabular}{c|cc|c|c|cc}
\hline
& & & & & \multicolumn{2}{c}{\textbf{Maximal substrings}} \\
\textbf{\CST}      & \textbf{\CSA} & \textbf{\LCP} & \textbf{Size} & \textbf{Traversal} & \textbf{Forward} & \textbf{Backward} \\
\hline
\CSTsada           & \CSAsada      & \PLCP         &     12.33 bpc &          0.06 \mus &       79.97 \mus &         5.14 \mus \\
\CSTnpr            & \SSA          & \PLCP         &     10.79 bpc &          0.23 \mus &       44.55 \mus &         0.46 \mus \\
\CSTnpr            & \SSA          & \LCPdac       &     18.08 bpc &          0.23 \mus &       29.70 \mus &         0.40 \mus \\
\FCST              & \SSA          & --            &      4.98 bpc &      (317.30 \mus) &      332.80 \mus &         3.13 \mus \\
\RCST               & \RFM          & \RLCP         &      2.75 bpc &          0.90 \mus &      208.62 \mus &         3.72 \mus \\
\RCST{} + \rselect  & \RFM          & \RLCP         &      3.21 bpc &          0.90 \mus &       80.20 \mus &         3.71 \mus \\
\hline
\end{tabular}
\end{center}
\end{table*}

It should be noted that the memory access patterns of traversing \CSTsada, \CSTnpr, and \RCST{} are highly local. Traversal times are mostly based on the amount of computation done, while memory latency is less important than in the individual query benchmarks. In \RCST{}, the algorithm is essentially the following: 1) compute \rmq{} in the current range; 2) proceed recursively to the left subinterval; and 3) proceed to the right subinterval. This involves plenty of redundant work, as can be seen by comparing the traversal time (0.90~\mus{} per node) to sequential \RLCP{} access (0.017~\mus{} per position). A faster algorithm would decompress large parts of the \LCP{} array at once, build the corresponding subtrees in postorder \cite{Abouelhoda2004}, and traverse the resulting trees.

\RCST{} with \rselect{} is as fast as \CSTsada{} in the forward algorithm,
1.8\nobreakdash--2.7~times slower than \CSTnpr, and 4.1~times faster than \FCST. Without
the additional structure, \RCST{} becomes 2.6~times slower. As expected \cite{Ohlebusch2010a},
the backward algorithm is much faster than the forward algorithm. \CSTsada{} and \RCST,
which combine slow backward searching with a fast tree, have similar
performance to \FCST, which combines fast searching with a slow tree. \CSTnpr{} is about
an order of magnitude faster than the others in the backward algorithm.
